src/Microsoft.SqlTools.ServiceLayer/Connection/ListDatabaseRequestHandler.cs

```diff
@@ -64,39 +65,74 @@ abstract class ListDatabaseRequestHandler<T> : IListDatabaseRequestHandler
 
         public ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo)
         {
+            ListDatabasesResponse response = new ListDatabasesResponse();
             ConnectionDetails connectionDetails = connectionInfo.ConnectionDetails.Clone();
+            // Running query against sys.databases view will only return a subset of databases the current login/user might have access to, we need to
+            // query the master database to get the full database list, but for users without master db access, we have to query the
+            // original database as a fallback.
+            var databasesToTry = new List<string>() { CommonConstants.MasterDatabaseName };
+            if (connectionDetails.DatabaseName != CommonConstants.MasterDatabaseName)
+            {
+                databasesToTry.Add(connectionDetails.DatabaseName);
+            }
+            for (int i = 0; i < databasesToTry.Count; i++)
+            {
+                try
+                {
+                    connectionDetails.DatabaseName = databasesToTry[i];
+                    var results = this.GetResults(connectionFactory, connectionDetails);
+                    SetResponse(response, results);
+                    break;
+                }
+                catch (Microsoft.Data.SqlClient.SqlException ex)
+                {
+                    // Retry when login attempt failed.
+                    // https://learn.microsoft.com/sql/relational-databases/errors-events/mssqlserver-18456-database-engine-error
+                    if (i != databasesToTry.Count - 1 && ex.Number == 18456)
```

Comment:
it may be worth logging a debug message here in case this results in any unexpected behavior.

Reply:
added logging.